### PR TITLE
terra node config

### PR DESCRIPTION
**Pull request by jmank88**
Add `terra_nodes` table and related CLI commands. Initialize terra tx manager with first node found in db.

### CODE REVIEW

**Comment by cl-homu** [comment]
:umbrella: The latest upstream changes (presumably 459dfa2b5cdf636ea4f4f94a07f0ccd6c031cae3) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

File: core/services/chainlink/application.go [hunk 1]
```diff
@@ -289,26 +294,37 @@ func NewApplication(opts ApplicationOpts) (Application, error) {
...
 	}
+	terraORM := terra.NewORM(db)
 	if cfg.FeatureOffchainReporting2() {
 		globalLogger.Debug("Off-chain reporting v2 enabled")
 		// TODO: parameterize poll period
-		// TODO: pull client config from nodes
-		// TODO: If terra enabled
-		chainID := "42"
-		tc, err := terraclient.NewClient("42", "0.01", "1.5", "TODO", "TODO", 10, globalLogger)
+		nodes, _, err := terraORM.Nodes(0, 1)
```
Comment (jmank88) [inline core/services/chainlink/application.go:301]: Temporary caveat: we are initializing the tx manager on startup, so after adding a node via remote CLI command, a restart is required to use the node.

File: core/cmd/terra_node_commands.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+	return nil
+}
+
+// TerraNodePresenters implements TableRenderer for a slice of TerraNodePresenter.
+type TerraNodePresenters []TerraNodePresenter
+
+// RenderTable implements TableRenderer
+func (ps TerraNodePresenters) RenderTable(rt RendererTable) error {
+	headers := []string{"ID", "Name", "Chain ID", "Tendermint URL", "FCD URL", "Created", "Updated"}
```
Comment (pinebit) [inline core/cmd/terra_node_commands.go:51]: would be great to have `headers` as a global variable, to avoid duplication.

**Review by pinebit — APPROVED** [review]

File: core/store/migrate/migrations/0092_terra_txm.sql [hunk 3]
```diff
@@ -1,5 +1,22 @@
 -- +goose Up
 -- +goose StatementBegin
+ALTER TABLE heads RENAME TO evm_heads;
+ALTER TABLE nodes RENAME TO evm_nodes;
+-- CREATE TABLE terra_chains (
```
Comment (connorwstein) [inline core/store/migrate/migrations/0092_terra_txm.sql:5]: I think we do want this table, is enabling it planned for a follow up PR?

File: core/store/migrate/migrations/0092_terra_txm.sql [hunk 4]
```diff
@@ -1,5 +1,22 @@
 -- +goose Up
 -- +goose StatementBegin
+ALTER TABLE heads RENAME TO evm_heads;
+ALTER TABLE nodes RENAME TO evm_nodes;
```
Comment (connorwstein) [inline core/store/migrate/migrations/0092_terra_txm.sql:4]: we could rename the eth_XXX tables to evm_XXX too I think cc @samsondav

**Review by connorwstein — APPROVED** [review]
some intermediate directories might help regarding organization i.e. 
cmd/nodes
web/presenters/chains/
web/presenters/nodes
web/nodes
web/chains

File: core/services/chainlink/application.go (hunk 1 above)
Comment (samsondav) [reply core/services/chainlink/application.go:301]: This applies to all chain/node changes currently, there is a ticket in the backlog to make it do hot loading https://app.shortcut.com/chainlinklabs/story/24556/need-to-be-able-to-boot-with-no-eth-chains-eth-nodes-and-add-them-later-and-have-everything-spin-up-as-expected

**Review by samsondav — COMMENTED** [review]
LGTM boring and no surprises, just how I like it :D

File: core/store/migrate/migrations/0092_terra_txm.sql (hunk 3 above)
Comment (jmank88) [reply core/store/migrate/migrations/0092_terra_txm.sql:5]: Yeah was just getting ahead of myself.

File: core/store/migrate/migrations/0092_terra_txm.sql [hunk 5]
```diff
@@ -1,5 +1,22 @@
...
+-- CREATE TABLE terra_chains (
+--     id text PRIMARY KEY,
+--     cfg jsonb NOT NULL DEFAULT '{}',
+--     created_at timestamptz NOT NULL,
+--     updated_at timestamptz NOT NULL
+-- );
```
[the remark this answers is not in the document]
Comment (jmank88) [reply core/store/migrate/migrations/0092_terra_txm.sql:10]: Just saving for upcoming PR. Not used yet.

File: core/store/migrate/migrations/0092_terra_txm.sql [hunk 6]
```diff
@@ -1,5 +1,22 @@
...
+CREATE TABLE terra_nodes (
+    id serial PRIMARY KEY,
+    name varchar(255) NOT NULL CHECK (name != ''),
```
[the remark this answers is not in the document]
Comment (jmank88) [reply core/store/migrate/migrations/0092_terra_txm.sql:13]: This is based on the existing `nodes` table, where we name evm nodes.

File: core/store/migrate/migrations/0092_terra_txm.sql [hunk 7]
```diff
@@ -1,5 +1,22 @@
...
+ 	terra_chain_id text NOT NULL, -- REFERENCES terra_chains (id),
```
[the remark this answers is not in the document]
Comment (jmank88) [reply core/store/migrate/migrations/0092_terra_txm.sql:14]: I think we'll want to support multiple nodes here to have a pool of clients just like evm, no?

**Comment by jmank88** [comment]
> some intermediate directories might help regarding organization i.e. cmd/nodes web/presenters/chains/ web/presenters/nodes web/nodes web/chains

Unfortunately, all of those seemingly trivial file moves require complex refactoring to resolve circular dependencies (more than just tests helpers too).

File: core/chains/terra/orm.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+// ErrNoRowsAffected is returned when rows should have been affected but were not.
+var ErrNoRowsAffected = errors.New("no rows affected")
+
+func (o orm) CreateNode(data types.NewNode) (node types.Node, err error) {
+	sql := `INSERT INTO terra_nodes (name, terra_chain_id, tendermint_url, fcd_url, created_at, updated_at)
+	VALUES (:name, :terra_chain_id, :tendermint_url, :fcd_url, now(), now())
+	RETURNING *;`
+	stmt, err := o.db.PrepareNamed(sql)
```
[the remark this answers is not in the document]
Comment (jmank88) [reply core/chains/terra/orm.go:29]: This was copied from `evm/orm.go` originally. Should I update that one as well?

File: core/store/migrate/migrations/0092_terra_txm.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (samsondav) [reply core/store/migrate/migrations/0092_terra_txm.sql]: Again this is me misreading the table, yes this looks fine

**Pull request merged**